### PR TITLE
Improved algebraic trait impls.

### PR DESCRIPTION
**Pull request by ryzhyk**
See commit messages.

### CODE REVIEW

File: src/algebra/finite_map.rs [hunk 1]
```diff
@@ -317,39 +317,102 @@ where
...
+                let entry = self.value.entry(k);
+                match entry {
+                    Entry::Vacant(e) => {
+                        e.insert(v);
+                    }
+                    Entry::Occupied(mut e) => {
+                        e.get_mut().add_assign(v);
+                        if e.get().is_zero() {
+                            e.remove_entry();
+                        };
```
Comment (Kixiron) [inline src/algebra/finite_map.rs:340]:
```suggestion
                        }
```

File: src/algebra/finite_map.rs [hunk 2]
```diff
@@ -317,39 +317,102 @@ where
...
+                let entry = other.value.entry(k);
+                match entry {
+                    Entry::Vacant(e) => {
+                        e.insert(v);
+                    }
+                    Entry::Occupied(mut e) => {
+                        e.get_mut().add_assign(v);
+                        if e.get().is_zero() {
+                            e.remove_entry();
+                        };
```
Comment (Kixiron) [inline src/algebra/finite_map.rs:356]:
```suggestion
                        }
```

File: src/algebra/finite_map.rs [hunk 3]
```diff
@@ -317,39 +317,102 @@ where
...
+                let entry = result.value.entry(k.clone());
+                match entry {
+                    Entry::Vacant(e) => {
+                        e.insert(v.clone());
+                    }
+                    Entry::Occupied(mut e) => {
+                        e.get_mut().add_assign_by_ref(v);
+                        if e.get().is_zero() {
+                            e.remove_entry();
+                        };
```
Comment (Kixiron) [inline src/algebra/finite_map.rs:384]:
```suggestion
                        }
```

File: src/algebra/finite_map.rs [hunk 4]
```diff
@@ -317,39 +317,102 @@ where
...
```
Comment (Kixiron) [inline src/algebra/finite_map.rs:403]:
```suggestion
                        }
```

**Review by Kixiron — COMMENTED** [review]
I think the arms of the two sides could be factored out into functions where you just reverse the order of the arguments, that'd be nice in removing some of the duplicate code.
There's also some more if statements with semicolons around the code you changed that GitHub wouldn't let me suggest on if you feel like fixing those up.
I'm also not sure I like the `.support_size()` function name, it's not immediately apparent as to what it means

**Comment by ryzhyk** [comment]
> I think the arms of the two sides could be factored out into functions where you just reverse the order of the arguments, that'd be nice in removing some of the duplicate code. There's also some more if statements with semicolons around the code you changed that GitHub wouldn't let me suggest on if you feel like fixing those up. I'm also not sure I like the `.support_size()` function name, it's not immediately apparent as to what it means

Thanks, fixed these issues and added some more optimizations.

File: src/algebra/finite_map.rs [hunk 5]
```diff
@@ -317,35 +315,109 @@ where
...
+                            e.remove_entry();
+                        }
+                    }
+                }
+            }
+            this
+        }
+
+        if self.support_size() > other.support_size() {
+            do_add(self.clone(), other)
```
Comment (mbudiu-vmw) [inline src/algebra/finite_map.rs:392]: both these do the same amount of work algorithmically. this optimization is justified only if clone is much faster than inserting all entries.

**Review by mbudiu-vmw — APPROVED** [review]

File: src/algebra/finite_map.rs (hunk 5 above)
Comment (ryzhyk) [reply src/algebra/finite_map.rs:392]: That's right, cloning the map should be more efficient than element-by-element insertion.

**Pull request merged**